text/0000-geoip-plugins.md

```diff
@@ -0,0 +1,140 @@
+can be used to geolocate the user based on their IP address and provide IP
+geolocation data to React components respectively.
+
+## Basic example
+
+### fusion-plugin-geoip
+
+```js
+const geoipData = GeoIP.from(ctx);
+console.log(geoipData.countryCode);
+console.log(geoipData.lat);
+console.log(geoipData.lng);
+```
+
+### fusion-plugin-geoip-react
+
+```js
+import {withGeoip} from 'fusion-plugin-geoip-react';
+
+function HelloGeoip({geoip}) {
+  return <p>Hello {geoip.countryName}</p>;
+}
+
+const HelloGeoipComponent = withGeoip(HelloGeoip);
+```
+
+## Motivation
+
+These plugins will support GeoIP, which is a common requirement for websites.
+For instance, GeoIP could be used to determine whether to show an EU cookie
+banner. Other potential use cases include CMS sites showing different content
+based on the user's country or map applications wanting to show a relevant
+section of the world to each user.
+
+The end goal is to have plugins that site builders can install, configure with
+a Maxmind database file, and immediately start using geoip in their Fusion
+application.
+	
+## Detailed design
+
+### fusion-plugin-geoip
+
+We will use `@lxe/maxmind-db-reader` to load the geoip data from `ctx.ip` on
+the server. A `geoip data` is a specific object that we define here using a
+Flow type. The Maxmind DB reader needs to be initialized with a Maxmind .mmdb
+file, so we additionally define a `GeoipConfigToken`. The types are as follows:
+
+```js
+type GeoipConfig = {
+  dbPath: string,
+  defaultGeoData?: GeoipData
+};
+
+type GeoipData = {
+  continentCode: string,
+  continentName: string,
+  countryCode: string,
+  countryName: string,
+  cityName: ?string,
+  lat: number,
+  lng: number,
+  timeZone: string, // e.g., "America/Los_Angeles"
+};
+```
+
+The type of `GeoIP.from` is `(ctx: Context) => GeoipData` and it will be
+memoized. The `dbPath` is required and the server should not start if it is not
+a valid path to a .mmdb database file. All consumers use `GeoIP.from` to load
+geoip data from the context. If geoip data is not available, then the user has
+the option to configure `defaultGeoData` that will be used instead. If geoip
+data is not available and the user does not configure `defaultGeoData`, geoip
+will resolve to a default value chosen by the package. This is acceptable
+because GeoIP is generally understood to be a best-effort estimation and can
+never achieve 100% accuracy.
+
+We will additionally define a middleware that includes the geoip data in JSON
+in a script tag on the page. We will then load this tag in the browser in
+order to provide geoip there. This design is similar to how
+fusion-plugin-i18n works.
+
+### fusion-plugin-geoip-react
+
+We will set up a context provider that provides GeoIP data, and a `withGeoip`
+HOC that consumes the geoip data from React context. The HOC will inject the
+`geoip` prop containing geoip data.
+
+## Drawbacks
```

Comment:
Any GDPR concerns that should be mentioned?

Reply:
I think most of the GDPR rules generally apply to data at rest, i.e., logs, customer records etc. It should be OK to use this data right away in a live website, but it might be problematic to generate logs of who was visiting from where. But if a site logs IP addresses, they could run geoip offline later anyway.